File: src/qb2api/checkin/scheduler.py
"""In-process check-in scheduler: zoneinfo, catch-up, single task (design §11)."""

from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Any
from zoneinfo import ZoneInfo

from qb2api.checkin.service import (
    CheckinInProgressError,
    CheckinService,
    jitter_seconds,
    next_run_after,
    parse_checkin_at,
)
from qb2api.config import Settings

logger = logging.getLogger("qb2api.checkin.scheduler")


@dataclass(frozen=True)
class ScheduleConfiguration:
    enabled: bool
    hour: int
    minute: int
    timezone: ZoneInfo
    catch_up: bool
    catch_up_window_hours: int
    jitter_min_seconds: int
    jitter_max_seconds: int

    @classmethod
    def from_settings(
        cls, settings: Settings, overrides: dict[str, Any] | None = None
    ) -> ScheduleConfiguration:
        values = overrides or {}
        checkin_at = str(values.get("checkin_at", settings.checkin_at))
        timezone = str(values.get("checkin_timezone", settings.checkin_timezone))
        hour, minute = parse_checkin_at(checkin_at)
        jitter_min = int(values.get("checkin_jitter_min_seconds", settings.checkin_jitter_min_seconds))
        jitter_max = int(values.get("checkin_jitter_max_seconds", settings.checkin_jitter_max_seconds))
        if jitter_min < 0 or jitter_max < jitter_min:
            raise ValueError("invalid checkin jitter range")
        window = int(values.get("checkin_catch_up_window_hours", settings.checkin_catch_up_window_hours))
        if window < 0:
            raise ValueError("invalid checkin catch-up window")
        return cls(
            enabled=bool(values.get("checkin_enabled", settings.checkin_enabled)),
            hour=hour, minute=minute, timezone=ZoneInfo(timezone),
            catch_up=bool(values.get("checkin_catch_up", settings.checkin_catch_up)),
            catch_up_window_hours=window, jitter_min_seconds=jitter_min, jitter_max_seconds=jitter_max,
        )


class CheckinScheduler:
    """One asyncio task; shared lock via CheckinService.is_running."""

    def __init__(self, service: CheckinService, settings: Settings) -> None:
        self._service = service
        self._settings = settings
        self._task: asyncio.Task | None = None
        self._stopped = asyncio.Event()
        self._configuration = ScheduleConfiguration.from_settings(settings)
        self._reconfigure_lock = asyncio.Lock()
        self._catch_up_decision = "not_evaluated"
        self._last_error: str | None = None
        self._last_run_at: datetime | None = None

    @property
    def next_run_at(self) -> datetime | None:
        if not self._configuration.enabled:
            return None
        return next_run_after(
            datetime.now(self._configuration.timezone),
            self._configuration.hour,
            self._configuration.minute,
        )

    def status_snapshot(self) -> dict[str, Any]:
        next_run = self.next_run_at
        return {
            "next_run_at": next_run.isoformat() if next_run else None,
            "catch_up_decision": self._catch_up_decision,
            "active_run_id": self._service.active_run_id,
            "last_error": self._last_error,
            "last_run_at": self._last_run_at.isoformat() if self._last_run_at else None,
        }

    def start(self) -> None:
        if not self._configuration.enabled:
            logger.info("checkin scheduler disabled")
            return
        if self._task is not None:
            return
        self._stopped.clear()
        self._task = asyncio.create_task(self._loop(), name="checkin-scheduler")
        logger.info(
            "checkin scheduler started at %s %s (tz=%s)",
            self._settings.checkin_at,
            self._settings.checkin_timezone,
            self._configuration.timezone,
        )

    async def stop(self) -> None:
        self._stopped.set()
        task = self._task
        self._task = None
        if task is not None:
            task.cancel()
            try:
                await task
            except asyncio.CancelledError:
                pass

    async def reconfigure(self, overrides: dict[str, Any] | None = None) -> None:
        """Validate the replacement schedule before cancelling the old timer."""
        candidate = ScheduleConfiguration.from_settings(self._settings, overrides)
        async with self._reconfigure_lock:
            await self.stop()
            self._configuration = candidate
            self._catch_up_decision = "not_evaluated"
            self._last_error = None
            self.start()

    async def _loop(self) -> None:
        try:
            await self._maybe_catch_up()
        except Exception as exc:
            self._last_error = type(exc).__name__
            logger.warning("checkin catch-up failed: %s", type(exc).__name__)

        while not self._stopped.is_set():
            now = datetime.now(self._configuration.timezone)
            nxt = next_run_after(now, self._configuration.hour, self._configuration.minute)
            delay = max(0.5, (nxt - now).total_seconds())
            try:
                await asyncio.wait_for(self._stopped.wait(), timeout=delay)
                break
            except TimeoutError:
                pass
            if self._stopped.is_set():
                break
            try:
                await self._launch_batch("scheduler")
            except CheckinInProgressError:
                self._last_error = "checkin_run_in_progress"
                logger.info("scheduled checkin skipped: already running")
            except Exception as exc:
                self._last_error = type(exc).__name__
                logger.warning("scheduled checkin failed: %s", type(exc).__name__)

    async def _maybe_catch_up(self) -> None:
        configuration = self._configuration
        if not configuration.catch_up:
            self._catch_up_decision = "disabled"
            return
        now = datetime.now(configuration.timezone)
        planned = now.replace(
            hour=configuration.hour, minute=configuration.minute, second=0, microsecond=0
        )
        if now < planned:
            self._catch_up_decision = "before_schedule"
            return
        window = timedelta(hours=configuration.catch_up_window_hours)
        if now - planned > window:
            self._catch_up_decision = "outside_window"
            logger.info("checkin catch-up skipped: outside window")
            return
        if not await self._has_pending_targets():
            self._catch_up_decision = "already_complete"
            logger.info("checkin catch-up skipped: all targets already terminal")
            return
        j = jitter_seconds(
            configuration.jitter_min_seconds,
            configuration.jitter_max_seconds,
        )
        if j > 0:
            await asyncio.sleep(j)
        if self._stopped.is_set():
            self._catch_up_decision = "stopped"
            return
        await self._launch_batch("catch_up")
        self._catch_up_decision = "started"

    async def _has_pending_targets(self) -> bool:
        checker = getattr(self._service, "has_pending_targets", None)
        if checker is None:
            return True
        return bool(await checker())

    async def _launch_batch(self, trigger: str) -> None:
        start_batch = getattr(self._service, "start_batch", None)
        if start_batch is None:
            await self._service.run_batch(trigger=trigger, skip_already_done=True)
        else:
            await start_batch(trigger=trigger, skip_already_done=True)
        self._last_run_at = datetime.now(self._configuration.timezone)

File: gobackend/internal/checkin/scheduler.go
package checkin

import (
	"context"
	"log/slog"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/dmego/qoderbuddy2api/gobackend/internal/store"
)

// Scheduler owns the daily timer and the catch-up decision.
type Scheduler struct {
	options SchedulerOptions

	mu              sync.Mutex
	stopped         bool
	done            chan struct{}
	lastRunAt       string
	lastError       string
	catchUpDecision string
}

// NewScheduler builds the scheduler.
func NewScheduler(options SchedulerOptions) *Scheduler {
	return &Scheduler{options: options, catchUpDecision: "not_evaluated"}
}

// Start launches the scheduler loop. A disabled schedule is a no-op, which keeps
// the status view honest rather than reporting a timer that never fires.
func (s *Scheduler) Start() {
	s.mu.Lock()
	if s.done != nil || s.stopped {
		s.mu.Unlock()
		return
	}
	if !s.enabled() {
		s.mu.Unlock()
		slog.Info("checkin scheduler disabled")
		return
	}
	s.done = make(chan struct{})
	done := s.done
	s.mu.Unlock()
	go s.loop(done)
}

// Stop halts the scheduler loop.
func (s *Scheduler) Stop() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.stopped = true
	if s.done != nil {
		close(s.done)
		s.done = nil
	}
}

// Reconfigure restarts the timer with the current settings.
func (s *Scheduler) Reconfigure() {
	s.mu.Lock()
	previous := s.done
	s.done = nil
	s.stopped = false
	s.catchUpDecision = "not_evaluated"
	s.lastError = ""
	s.mu.Unlock()
	if previous != nil {
		close(previous)
	}
	s.Start()
}

// loop runs catch-up once before sleeping, so a restart repairs a missed run
// immediately instead of waiting a full day.
func (s *Scheduler) loop(done chan struct{}) {
	s.maybeCatchUp(done)
	for {
		delay := s.durationUntilNextRun()
		timer := time.NewTimer(delay)
		select {
		case <-done:
			timer.Stop()
			return
		case <-timer.C:
		}
		s.launch("scheduler")
	}
}

// durationUntilNextRun is the time to the next scheduled instant, with a floor
// so a schedule landing exactly on "now" still sleeps rather than spinning.
func (s *Scheduler) durationUntilNextRun() time.Duration {
	location, err := time.LoadLocation(s.timezone())
	if err != nil {
		location = time.UTC
	}
	hour, minute := parseClock(s.checkinAt())
	now := time.Now().In(location)
	next := time.Date(now.Year(), now.Month(), now.Day(), hour, minute, 0, 0, location)
	// Inclusive boundary: a candidate equal to now rolls to tomorrow.
	if !next.After(now) {
		next = next.AddDate(0, 0, 1)
	}
	delay := time.Until(next)
	if delay < 500*time.Millisecond {
		return 500 * time.Millisecond
	}
	return delay
}

// maybeCatchUp decides whether today's scheduled run was missed.
func (s *Scheduler) maybeCatchUp(done chan struct{}) {
	if !s.catchUpEnabled() {
		s.setDecision("disabled")
		return
	}
	location, err := time.LoadLocation(s.timezone())
	if err != nil {
		location = time.UTC
	}
	hour, minute := parseClock(s.checkinAt())
	now := time.Now().In(location)
	planned := time.Date(now.Year(), now.Month(), now.Day(), hour, minute, 0, 0, location)
	if now.Before(planned) {
		s.setDecision("before_schedule")
		return
	}
	window := time.Duration(s.catchUpWindowHours()) * time.Hour
	if now.Sub(planned) > window {
		s.setDecision("outside_window")
		slog.Info("checkin catch-up skipped: outside window")
		return
	}
	if !s.hasPendingTargets() {
		s.setDecision("already_complete")
		slog.Info("checkin catch-up skipped: all targets already terminal")
		return
	}
	jitter := s.jitterDuration()
	if jitter > 0 {
		select {
		case <-done:
			s.setDecision("stopped")
			return
		case <-time.After(jitter):
		}
	}
	s.setDecision("started")
	s.launch("catch_up")
}

// hasPendingTargets reports whether any eligible account still needs today's run.
func (s *Scheduler) hasPendingTargets() bool {
	if s.options.Service == nil || s.options.Service.options.Registry == nil {
		return false
	}
	targets := s.options.Service.options.Registry.EligibleForPurpose("codebuddy", "checkin")
	if len(targets) == 0 {
		return false
	}
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	for _, target := range targets {
		state, err := s.options.Service.options.DB.GetCheckinDailyState(ctx, target.Provider, target.AccountID,
			s.options.Service.localDate(), s.timezone())
		if err != nil || state.TerminalOutcome == nil {
			return true
		}
		outcome := *state.TerminalOutcome
		if outcome != OutcomeClaimed && outcome != OutcomeAlreadyCheckedIn {
			return true
		}
	}
	return false
}

// launch starts a batch, swallowing the in-progress error because a concurrent
// run already covers the same accounts.
func (s *Scheduler) launch(trigger string) {
	if s.options.Service == nil {
		return
	}
	skipAlreadyDone := trigger == "scheduler" || trigger == "catch_up"
	runID, err := s.options.Service.StartBatch(context.Background(), trigger, nil, skipAlreadyDone)
	s.mu.Lock()
	defer s.mu.Unlock()
	if err != nil {
		if err == ErrRunInProgress {
			s.lastError = "checkin_run_in_progress"
			slog.Info("checkin run already in progress")
			return
		}
		s.lastError = typeName(err)
		slog.Warn("checkin batch failed to start", "error", err)
		return
	}
	s.lastError = ""
	s.lastRunAt = store.FormatISO(time.Now())
	slog.Info("checkin batch started", "trigger", trigger, "run_id", runID)
}

// Status renders the scheduler snapshot.
func (s *Scheduler) Status() map[string]any {
	s.mu.Lock()
	defer s.mu.Unlock()
	status := map[string]any{
		"catch_up_decision": s.catchUpDecision,
		"active_run_id":     nil,
		"last_error":        nilable(s.lastError),
		"last_run_at":       nilable(s.lastRunAt),
		"next_run_at":       nil,
	}
	if s.options.Service != nil {
		status["active_run_id"] = nilable(s.options.Service.ActiveRunID())
	}
	if s.enabled() {
		location, err := time.LoadLocation(s.timezone())
		if err != nil {
			location = time.UTC
		}
		hour, minute := parseClock(s.checkinAt())
		now := time.Now().In(location)
		next := time.Date(now.Year(), now.Month(), now.Day(), hour, minute, 0, 0, location)
		if !next.After(now) {
			next = next.AddDate(0, 0, 1)
		}
		status["next_run_at"] = next.Format(time.RFC3339)
	}
	return status
}

func (s *Scheduler) setDecision(value string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.catchUpDecision = value
}

func (s *Scheduler) enabled() bool {
	if s.options.Runtime != nil {
		return s.options.Runtime.CheckinEnabled()
	}
	return s.options.Settings.CheckinEnabled
}

func (s *Scheduler) checkinAt() string {
	if s.options.Runtime != nil {
		return s.options.Runtime.CheckinAt()
	}
	return s.options.Settings.CheckinAt
}

func (s *Scheduler) timezone() string {
	if s.options.Runtime != nil {
		return s.options.Runtime.CheckinTimezone()
	}
	return s.options.Settings.CheckinTimezone
}

func (s *Scheduler) catchUpEnabled() bool {
	if s.options.Runtime != nil {
		return s.options.Runtime.CheckinCatchUp()
	}
	return s.options.Settings.CheckinCatchUp
}

func (s *Scheduler) catchUpWindowHours() int {
	if s.options.Runtime != nil {
		return s.options.Runtime.CheckinCatchUpWindowHours()
	}
	return s.options.Settings.CheckinCatchUpHours
}

func (s *Scheduler) jitterDuration() time.Duration {
	minimum, maximum := s.options.Settings.CheckinJitterMin, s.options.Settings.CheckinJitterMax
	if s.options.Runtime != nil {
		minimum = s.options.Runtime.CheckinJitterMinSeconds()
		maximum = s.options.Runtime.CheckinJitterMaxSeconds()
	}
	lower, upper := minimum, maximum
	if lower > upper {
		lower, upper = upper, lower
	}
	if upper <= 0 {
		return 0
	}
	return time.Duration(lower+randomInt(upper-lower+1)) * time.Second
}

// defaultCheckinHour/Minute is the documented fallback schedule (00:10).
const (
	defaultCheckinHour   = 0
	defaultCheckinMinute = 10
)

// parseClock reads an HH:MM schedule time.
//
// Malformed input falls back to the documented default rather than 00:00: a
// typo in CHECKIN_AT should not silently move the daily run to midnight, and
// returning a fixed value keeps the scheduler's behaviour predictable.
func parseClock(value string) (int, int) {
	parts := strings.Split(strings.TrimSpace(value), ":")
	if len(parts) != 2 {
		return defaultCheckinHour, defaultCheckinMinute
	}
	hour, err := strconv.Atoi(strings.TrimSpace(parts[0]))
	if err != nil || hour < 0 || hour > 23 {
		return defaultCheckinHour, defaultCheckinMinute
	}
	minute, err := strconv.Atoi(strings.TrimSpace(parts[1]))
	if err != nil || minute < 0 || minute > 59 {
		return defaultCheckinHour, defaultCheckinMinute
	}
	return hour, minute
}
